lsuio.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" lsuio is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH LSUIO 1 "April 17, 2011"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
lsuio \- list all UIO devices.
.SH SYNOPSIS
.B lsuio
.RI [ options ]
.SH DESCRIPTION
\fBlsuio\fP lists all UIO devices in your system.
.PP
.SH OPTIONS
\fBlsuio\fP accepts the following options:
.TP
.BR  -a ", " --access
open and close all UIO devices
.TP
.BR  -v ", " --verbose
Print more information.
.TP
.BR  -? ", " --help
List all available options and their meanings.
.TP
.B  --usage
Show summary of options.
.TP
.BR  -V ", " --version
Show version of program.
.\" .SH "SEE ALSO"
.\" .BR foo (1), 
.\" .BR bar (1).
.SH AUTHOR
Hans-Jürgen Koch.


debian/changelog
libuio (0.2.8) unstable; urgency=medium

  * fix debianization and Makefile.am
  * irq: fix timeout handling
  * fix includes for use with newer glibc
  * add error handling for mmap

 -- Manuel Traut <manut@linutronix.de>  Thu, 03 May 2018 17:59:37 +0200

libuio (0.2.7) unstable; urgency=medium

  * Fix initial package build
  * Make package multiarch aware

 -- Benedikt Spranger <b.spranger@linutronix.de>  Mon, 27 Jul 2015 15:17:06 +0200

libuio (0.2.6-1) unstable; urgency=low

  * some fixes
  * new accessor function for uio file descriptor
  * rename .texinfo files

 -- Benedikt Spranger <b.spranger@linutronix.de>  Wed, 28 Aug 2013 11:50:42 +0200

libuio (0.2.5-1) unstable; urgency=low

  * add some conviniant functions
  * some bugfixes
  * better handling of attributes

 -- Benedikt Spranger <b.spranger@linutronix.de>  Wed, 28 Aug 2013 11:48:28 +0200

libuio (0.2.4-1) unstable; urgency=low

  * add function uio_find_by_uio_name()

 -- Benedikt Spranger <b.spranger@linutronix.de>  Wed, 12 Sep 2012 12:18:18 +0200

libuio (0.2.3-1) unstable; urgency=low

  * various fixes
  * merge debian support into upstrem

 -- Benedikt Spranger <b.spranger@linutronix.de>  Mon, 18 Jun 2012 21:12:58 +0200

libuio (0.2.2-1) unstable; urgency=low

  * attribute helper
  * i18n support
  * 64bit access macros
  * various fixes

 -- Benedikt Spranger <b.spranger@linutronix.de>  Wed, 01 Feb 2012 11:14:39 +0100

libuio (0.2.1-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- Benedikt Spranger <b.spranger@linutronix.de>  Thu, 14 Apr 2011 19:19:14 +0200


lsuio.c
/*
 * lsuio - libuio example application.
 *
 * Copyright (C) 2011 Benedikt Spranger
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */

#include <argp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/types.h>

#include "config.h"
#include "libuio.h"

#ifdef USE_GLIB
#include <glib.h>
#else
#define g_print	printf
#endif

#define EXIT_FAILURE 1

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# define textdomain(Domain)
# define _(Text) Text
#endif
#define N_(Text) Text

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static void show_version (FILE *stream, struct argp_state *state);

/* argp option keys */
enum {DUMMY_KEY = 129};

/* Option flags and variables.  These are initialized in parse_opt.  */

int want_verbose;		/* --verbose */
int want_access;		/* --access */

static struct argp_option options [] =
{
	{"access", 'a', NULL, 0, N_("open and close all devices"), 0},
	{"verbose", 'v', NULL, 0, N_("Print more information"), 0},
	{NULL, 0, NULL, 0, NULL, 0}
};

/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "https://github.com/linutronix/libuio/issues";
void (*argp_program_version_hook) (FILE *, struct argp_state *) = show_version;

static struct argp argp =
{
	options, parse_opt, NULL,
	N_("list uio devices."),
	NULL, NULL, NULL
};

int main (int argc, char **argv)
{
	struct uio_info_t **uio_list, *info;
	char **attr, **tmp;
	int i;

	textdomain (PACKAGE);
	argp_parse (&argp, argc, argv, 0, NULL, NULL);

	uio_list = uio_find_devices ();
	if (!uio_list)
	{
		g_print (_("No UIO devices found\n"));
		return 1;
	}

	for (;*uio_list; uio_list++)
	{
		info = *uio_list;
		g_print (_("Name   : %s\n"), uio_get_name (info));
		g_print (_("Version: %s\n"), uio_get_version (info));
		g_print (_("DevId  : %d:%d\n"), uio_get_major (info),
			uio_get_minor (info));
		g_print (_("DevNode: %s\n"), uio_get_devname (info));

		if (want_verbose)
		{
			g_print (_("Map    :\n"));
			for (i = 0; i < uio_get_maxmap (info); i++)
			{
				g_print (_("%3d addr: 0x%08lx\n"), i,
					uio_get_mem_addr (info, i));
				g_print (_("    size: 0x%08zx\n"),
					uio_get_mem_size (info, i));
				g_print (_("  offset: 0x%08zx\n"),
					uio_get_offset (info, i));
			}
		}
		g_print (_("\n"));

		g_print (_("Attr.  :\n"));
		attr = uio_list_attr (info);
		for (tmp = attr; tmp && *tmp; tmp++)
		{
			g_print (_("  %s\n"), *tmp);
			if (want_verbose)
				if (!strcmp (*tmp, "name") ||
				    !strcmp (*tmp, "version"))
					g_print (_("    %s\n"),
						uio_get_attr (info, *tmp));
			free (*tmp);
		}
		free (attr);

		if (want_access)
		{
			g_print (_("open : "));
			if (!uio_open (info))
				g_print (_("OK\n"));
			else
				g_print (_("failed\n"));

			g_print (_("close: "));
			if (!uio_close (info))
				g_print (_("OK\n"));
			else
				g_print (_("failed\n"));
		}
	}

	free (uio_list);

	return 0;
}

/* Parse a single option.  */
static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key)
	{
	case ARGP_KEY_INIT:
		/* Set up default values.  */
		want_verbose = 0;
		want_access = 0;
		break;

	case 'a':			/* --access */
		want_access = 1;
		break;

	case 'v':			/* --verbose */
		want_verbose = 1;
		break;

	case ARGP_KEY_ARG:		/* [FILE]... */
		/* TODO: Do something with ARG, or remove this case and make
         main give argp_parse a non-NULL fifth argument.  */
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

/* Show the version number and copyright information.  */
static void show_version (FILE *stream, struct argp_state *state)
{
	(void) state;

	fputs (PACKAGE" "VERSION"\n", stream);
	fprintf (stream, _("Written by %s.\n\n"), "Benedikt Spranger");
	fprintf (stream, _("Copyright (C) %s %s\n"), "2011", "Benedikt Spranger");
	fputs(_("\
This program is free software; you may redistribute it under the terms of\n\
the GNU General Public License.  This program has absolutely no warranty.\n"),
	      stream);
}
